### ISSUE TITLE
bdev_nvme/nvme: Race condition in nvme_ctrlr_destruct during hot remove

### ISSUE THREAD

**Issue by mhae** [spdk/spdk#2012]
Hot removing a large number of devices (~14) at the same time can lead to a SIGSEGV due to a race condition during freeing of the io qpairs.

Our SPDK target app uses NVMF and bdev_nvme to expose NVMe devices over NVMF. When we remove a large number of devices at the same time, we see a SIGSEGV during freeing of the ioqpairs:

```
#1  0x00007fcb619cfb25 in abort () from /lib64/libc.so.6
#2  0x000000000048d761 in el::base::utils::abort (status=11, reason="Application has crashed due to [SIGSEGV] signal - [Called el::Helpers::crashAbort(11)]") at external/easyloggingpp/easylogging++.cc:125
#3  el::Helpers::crashAbort (sig=11, sourceFile=0x59d89c "", line=0) at external/easyloggingpp/easylogging++.cc:3121
#4  <signal handler called>
#5  nvme_ctrlr_get_process (ctrlr=ctrlr@entry=0x0, pid=897074) at nvme_ctrlr.c:2314
#6  0x00000000004f03e1 in nvme_ctrlr_get_current_process (ctrlr=ctrlr@entry=0x0) at nvme_ctrlr.c:2326
#7  0x00000000004f087e in spdk_nvme_ctrlr_free_io_qpair (qpair=0x200067c1fe58) at nvme_ctrlr.c:598
#8  0x00000000004f1351 in nvme_ctrlr_destruct (ctrlr=ctrlr@entry=0x20005b054580) at nvme_ctrlr.c:2957
#9  0x000000000050105c in spdk_nvme_detach (ctrlr=0x20005b054580) at nvme.c:80
#10 0x00000000004c7e96 in nvme_bdev_unregister_cb (io_device=0x7fcb5842c8f0) at common.c:125
#11 0x0000000000515fa5 in _finish_unregister (arg=0x7fcb5842cc20) at thread.c:1220
#12 0x0000000000516a24 in msg_queue_run_batch (max_msgs=<optimized out>, thread=0x7fcb5889d060) at thread.c:505
#13 thread_poll (now=940964954168686, max_msgs=0, thread=0x7fcb5889d060) at thread.c:581
#14 spdk_thread_poll (thread=thread@entry=0x7fcb5889d060, max_msgs=max_msgs@entry=0, now=940964954168686) at thread.c:689
#15 0x0000000000532dd8 in _reactor_run (reactor=0x7fcb5889c9c0) at reactor.c:326
#16 reactor_run (arg=0x7fcb5889c9c0) at reactor.c:382
#17 0x000000000053332a in spdk_reactors_start () at reactor.c:477
#18 0x00000000005313e8 in spdk_app_start (opts=0x7fff8d2aa5d0, start_fn=start_fn@entry=0x42bcc0 <nvgrid::nvmf_tgt_started(void*)>, arg1=0x7fff8d2aa4bc) at app.c:710
```
The qpair 0x200067c1fe58 has already been freed by bdev_nvme_destroy_cb.

Here's a trace that shows the sequence leading to the abort. Note, I added additional tracing and our code is also slightly modified so the line numbers won't match stock 20.07. I also skipped some of the NVMF related trace messages.

```
// Device @c7.00.0 is hot removed
2021-06-26 11:22:03.971 MST {897074} E {nvmf_tgt_app} DPDK:0: EAL: Unexpected device scan at /sys/bus/pci/devices/0000:c7:00.0/driver!
2021-06-26 11:22:04.052 MST {897074} I {nvmf_tgt_app} nvme_transport.c:338:nvme_transport_ctrlr_disconnect_qpair Disconnecting ctrlr=0000:c7:00.0 cntlid=0x0021 qid=0x0000
2021-06-26 11:22:04.052 MST {897074} E {nvmf_tgt_app} nvme_pcie.c:1462:nvme_pcie_qpair_abort_trackers aborting outstanding command

2021-06-26 11:22:04.053 MST {897074} E {nvmf_tgt_app} nvme_ctrlr.c:825:nvme_ctrlr_fail ctrlr 0000:c7:00.0 in failed state.
2021-06-26 11:22:04.053 MST {897074} I {nvmf_tgt_app} bdev_nvme.c:1670:remove_cb remove_cb: 0000:c7:00.0
2021-06-26 11:22:04.054 MST {897074} I {nvmf_tgt_app} bdev_nvme.c:1687:remove_cb Removed ns: 0000:c7:00.0
2021-06-26 11:22:04.070 MST {897074} E {nvmf_tgt_app} DPDK:0: EAL: Unexpected device scan at /sys/bus/pci/devices/0000:c7:00.0/driver!
2021-06-26 11:22:04.090 MST {897074} I {nvmf_tgt_app} ctrlr.c:108:nvmf_ctrlr_stop_keep_alive_timer !!! Stopping ka for cntlid=0x0007

2021-06-26 11:22:04.091 MST {897074} I {nvmf_tgt_app} bdev_nvme.c:503:bdev_nvme_reset ??? ctrlr is in destruct: 0000:c7:00.0

// The bdev_nvme_destroy_cb is starting to free the qpairs. This is happening on all 4 cores that we have allocated for the target.
2021-06-26 11:22:04.097 MST {897074} I {nvmf_tgt_app} bdev_nvme.c:887:bdev_nvme_destroy_cb ??? calling spdk_nvme_ctrlr_free_io_qpair, qpair=0x200067c1fe58
// Entering spdk_nvme_ctrlr_free_io_qpair 
2021-06-26 11:22:04.097 MST {897074} I {nvmf_tgt_app} nvme_ctrlr.c:560:spdk_nvme_ctrlr_free_io_qpair ???? qpair=0x200067c1fe58
// spdk_nvme_ctrlr_free_io_qpair  is calling nvme_pcie_qpair_destroy 
2021-06-26 11:22:04.097 MST {897074} I {nvmf_tgt_app} nvme_pcie.c:1530:nvme_pcie_qpair_destroy ??? enter: qpair=0x200067c1fe58
2021-06-26 11:22:04.097 MST {897074} I {nvmf_tgt_app} bdev_nvme.c:887:bdev_nvme_destroy_cb ??? calling spdk_nvme_ctrlr_free_io_qpair, qpair=0x200066c27e58
2021-06-26 11:22:04.097 MST {897074} I {nvmf_tgt_app} nvme_ctrlr.c:560:spdk_nvme_ctrlr_free_io_qpair ???? qpair=0x200066c27e58
// This message is printed right after spdk_free(pqpair); 
2021-06-26 11:22:04.097 MST {897074} I {nvmf_tgt_app} nvme_pcie.c:1555:nvme_pcie_qpair_destroy ??? [nvmf_tgt_poll_group_2] exit (freed): qpair=0x200067c1fe58
2021-06-26 11:22:04.097 MST {897074} I {nvmf_tgt_app} bdev_nvme.c:887:bdev_nvme_destroy_cb ??? calling spdk_nvme_ctrlr_free_io_qpair, qpair=0x200066f3fe58
2021-06-26 11:22:04.097 MST {897074} I {nvmf_tgt_app} nvme_ctrlr.c:560:spdk_nvme_ctrlr_free_io_qpair ???? qpair=0x200066f3fe58
2021-06-26 11:22:04.098 MST {897074} I {nvmf_tgt_app} nvme_pcie.c:1530:nvme_pcie_qpair_destroy ??? enter: qpair=0x200066c27e58
2021-06-26 11:22:04.098 MST {897074} I {nvmf_tgt_app} nvme_pcie.c:1555:nvme_pcie_qpair_destroy ??? [nvmf_tgt_poll_group_4] exit (freed): qpair=0x200066c27e58
2021-06-26 11:22:04.098 MST {897074} I {nvmf_tgt_app} nvme_pcie.c:1530:nvme_pcie_qpair_destroy ??? enter: qpair=0x200066f3fe58

2021-06-26 11:22:04.113 MST {897074} I {nvmf_tgt_app} nvme_pcie.c:1555:nvme_pcie_qpair_destroy ??? [nvmf_tgt_poll_group_3] exit (freed): qpair=0x200066f3fe58
2021-06-26 11:22:04.113 MST {897074} I {nvmf_tgt_app} bdev_nvme.c:887:bdev_nvme_destroy_cb ??? calling spdk_nvme_ctrlr_free_io_qpair, qpair=0x200066d3fe58
2021-06-26 11:22:04.113 MST {897074} I {nvmf_tgt_app} nvme_ctrlr.c:560:spdk_nvme_ctrlr_free_io_qpair ???? qpair=0x200066d3fe58
2021-06-26 11:22:04.113 MST {897074} I {nvmf_tgt_app} bdev_nvme.c:503:bdev_nvme_reset ??? ctrlr is in destruct: 0000:c7:00.0
2021-06-26 11:22:04.113 MST {897074} I {nvmf_tgt_app} nvme_pcie.c:1530:nvme_pcie_qpair_destroy ??? enter: qpair=0x200066d3fe58

2021-06-26 11:22:04.130 MST {897074} I {nvmf_tgt_app} nvme_pcie.c:1555:nvme_pcie_qpair_destroy ??? [nvmf_tgt_poll_group_1] exit (freed): qpair=0x200066d3fe58

// At this point all 4 qpairs have been freed

2021-06-26 11:22:04.131 MST {897074} I {nvmf_tgt_app} bdev_nvme.c:503:bdev_nvme_reset ??? ctrlr is in destruct: 0000:c7:00.0

// Now, nvme_ctrlr_destruct is called (via nvme_bdev_unregister_cb), and it tries to walk through its list of ctrlr->active_io_qpairs.
// I'm assuming this is happening in parallel of the above cleanup.
// Since active_io_qpairs is unprotected, it seems to get an already freed qpair which leads the SIGSEGV.
2021-06-26 11:22:04.131 MST {897074} I {nvmf_tgt_app} nvme_ctrlr.c:2945:nvme_ctrlr_destruct ??? [app_thread] Prepare to destruct SSD: 0000:c7:00.0 (ctrlr=0x20005b054580), is_failed=1
2021-06-26 11:22:04.131 MST {897074} I {nvmf_tgt_app} nvme_ctrlr.c:560:spdk_nvme_ctrlr_free_io_qpair ???? qpair=0x200067c1fe58

CRASHED!
```

## Possible Solution
I wasn't sure what the best approach for dealing with this race would be. The ctrlr->active_io_qpairs should be protected, since all other mutations of active_io_qpairs are done under a ctrlr lock.
Another approach would be to delay the nvme_ctrlr_destruct until the cleanup via the bdev_nvme_destroy_cb has completed.

Any advice is appreciated.

## Context (Environment including OS version, SPDK version, etc.)
SPDK 20.07, CentOS8

**Comment by mhae** [comment]
I did some more testing. Protecting the active_io_qpairs iteration with the (recursive) ctrlr_lock does not cause the issue anymore:
```
void nvme_ctrlr_destruct(struct spdk_nvme_ctrlr *ctrlr)
...
// mhae
nvme_robust_mutex_lock(&ctrlr->ctrlr_lock);
TAILQ_FOREACH_SAFE(qpair, &ctrlr->active_io_qpairs, tailq, tmp) {
spdk_nvme_ctrlr_free_io_qpair(qpair);
}
// mhae
nvme_robust_mutex_unlock(&ctrlr->ctrlr_lock);
...
``` 
However, I don't think this is sufficient since qpair is accessed all over the place in spdk_nvme_ctrlr_free_io_qpair.

I wasn't able to fully grasp the cleanup. It seems that the bdev layer is putting back channels which results in (async) calls to `bdev_nvme_destroy_cb` for each channel.
The other path is via `nvme_bdev_ctrlr_destruct` which calls `spdk_io_device_unregister` with `nvme_bdev_unregister_cb`. The last function destructs the nvme ctrlr which also tries to free io qpairs.

**Comment by shuhei-matsumoto** [comment]
Hi @mhae 

The callback to spdk_io_device_unregister() is called after all IO channels are closed, i.e., nvme_bdev_unregister_cb() is called only after all bdev_nvme_destroy_cb() are returned.

So if spdk_nvme_ctrlr_free_io_qpair() completes freeing IO qpair synchronously, this issue should not occur?

I guess that spdk_nvme_ctrlr_free_io_qpair() returned by setting qpair->delete_after_completion_context to 1, and then the deferred call spdk_nvme_ctrlr_free_io_qpair() was raced with nvme_ctrlr_destruct().

We can add callback to spdk_nvme_ctrlr_free_io_qpair(). However, callback to spdk_put_io_channel() must be synchronous.

It will be nice if we can make both bdev_nvme_destroy_cb (the callback to spdk_put_io_channel'()) and spdk_nvme_ctrlr_free_io_qpair() asynchronous.

I will investigate how we can make the callback to spdk_put_io_channel() asynchronous if possible.

I think your workaround is good temporary fix.

-Shuhei

**Comment by shuhei-matsumoto** [comment]
It looks that we can make both spdk_nvme_ctrlr_free_io_qpair() and the callback to spdk_put_io_channel() asynchronous. But it will not meet your schedule, and so your workaround will be better. I will work on both in my multipath work but I'm fine if you do :)

**Comment by mhae** [comment]
@shuhei-matsumoto , thanks for the quick reply! The workaround I'm proposing in 'nvme_ctrlr_destruct' is not tight enough. I think the whole `spdk_nvme_ctrlr_free_io_qpair` function should be protected by the ctrlr lock. There should also be a check if the qpair is still in 'active_io_qpairs' at the beginning of the function and if it isn't, return immediately.
What do you think?

**Comment by shuhei-matsumoto** [comment]
@mhae Do you keep getting lock in nvme_ctrlr_destruct()? If so, I think your idea should work. I think eventually it is nice if ctrlr->active_io_qpairs is empty when nvme_ctrlr_destruct() is called.

**Comment by jimharris** [comment]
Hi @mhae,

Like you point out in the log, all 4 qpairs should be freed before nvme_ctrlr_destruct is called.  But that happens >10ms after the qpairs are freed earlier in the log.  In fact, the one that you crash on was supposedly freed >30ms before the crash.  Removing the qpairs from the active_io_qpairs list is protected, so the TAILQ can't be getting corrupted.  I agree that the TAILQ traversal needs to be protected by the lock, but it makes no sense yet why that makes a difference (although clearly it does by your report) since that traversal happens quite a while after the qpairs were removed from active_io_qpairs (and those removals are protected by the lock).

Regarding Shuhei's theory, I don't see how we could be deferring the qpair deletion - that only happens when we try to free the io qpair in the context of the completion handler - but your logs pretty clearly show that the io qpair is getting freed by the bdev_nvme_destroy_cb.  We also see 4 calls to nvme_pcie_qpair_destroy, indicating that each call to spdk_nvme_ctrlr_free_io_qpair() resulted in an actual qpair deletion.

Maybe in spdk_nvme_ctrlr_free_io_qpair(), after grabbing the lock and removing the qpair from active_io_qpairs, could you try printing out what's in active_io_qpairs?  Since that is all protected, we should see the qpairs disappear one by one as each core frees it qpair.  It might also help point out any unexpected TAILQ corruption.

-Jim

**Comment by shuhei-matsumoto** [comment]
@jimharris Thank you so much for your recovery. @mhae I'm sorry for my wrong estimation.

I noticed that spdk_nvme_ctrlr_free_io_qpair() can be called by nvme_ctrlr_cleanup_process() or nvme_ctrlr_remove_process(). Do you find any possibility that spdk_nvme_ctrlr_free_io_qpair() is called multiple times to the same qpair? But please focus on Jim's feedback.

**Comment by mhae** [comment]
@jimharris , @shuhei-matsumoto , thanks for your analysis... It actually was a TAILQ corruption.

The spdk_nvme_ctrlr_free_io_qpair function got called twice for one specific qpair. We issue a bdev_reset in some scenarios, which does attempt to free qpairs as well. This is happening before channel tear down and ctrlr destruct. 

Now, since the device is no longer there, the first call (from _bdev_nvme_reset_destroy_qpair) to spdk_nvme_ctrlr_free_io_qpair fails in nvme_transport_ctrlr_delete_io_qpair. Note that the qpair is removed from TAILQ before the call to nvme_transport_ctrlr_delete_io_qpair. 
When the channels are closed, the TAILQ remove for an already removed qpair is  executed again, potentially adding an already freed qpair back to the list.

So, I'm not 100% sure what to do here. I do think the spdk_nvme_ctrlr_free_io_qpair needs to be better protected since we have multiple paths into.
The safest would be to check if the qpair is still in the list. We could also move the  remove after the nvme_transport_ctrlr_delete_io_qpair but I'm not sure what the guarantees and error scenarios are the transport ctrlr_delete_io_qpair.

**Comment by jimharris** [comment]
Ah, that makes sense now @mhae.  This is a bug in the bdev/nvme module, and I think the fix will actually be pretty easy.

When we reset the controller, we iterate over all the channels (i.e. cores) and free each of the qpairs.  Once all of the qpairs are freed, we do the spdk_nvme_ctrlr_reset, and then iterate over all the channels again to allocate new qpairs.

But if this races with the destruct operation, the bdev_nvme_destroy_cb could free the qpair first.  _bdev_nvme_reset_destroy_qpair could then execute and free the same qpair.  It looks fairly easy to make these two paths be aware of each other and avoid the double free.

I'll get a patch for you here shortly @mhae.

**Comment by jimharris** [comment]
[patch.txt](https://github.com/spdk/spdk/files/6749217/patch.txt)

**Comment by mhae** [comment]
Sweet. We'll give it a try. I had the same idea originally but I didn't want to completely eliminate the error handling in spdk_nvme_ctrlr_free_io_qpair.
I do think however, that this may lead to a memory leak if the qpair can't be destructed (freed) in spdk_nvme_ctrlr_free_io_qpair. The actually free() of the qpair is happening  nvme_transport_ctrlr_delete_io_qpair. So if in the reset we fail to successfully execute nvme_transport_ctrlr_delete_io_qpair, then we will leak the qpair. In our case, since the device was gone, we the call to nvme_transport_ctrlr_delete_io_qpair failed.

**Comment by jimharris** [comment]
Good point @mhae.

I think we should just make the pcie transport always report success, even if it could not actually delete the SQ and CQ (i.e. when the controller is hot removed).  Realistically there's nothing the caller can do to recover from this condition.  Even if SQ/CQ deletion failed for some reason other than hotplug, the caller has asked that the nvme driver delete/free the IO qpair so it needs to honor that.

Note that TCP and RDMA never return failure for their delete_io_qpair implementations.  vfio_user shares the PCIe implementation.

I'll suggest a different patch here in a bit.

**Comment by jimharris** [comment]
[patch_pcie_no_failure.txt](https://github.com/spdk/spdk/files/6750212/patch_pcie_no_failure.txt)

Can you try this patch together with the earlier patch?

This patch here ensures nvme_pcie_ctrlr_delete_io_qpair() never returns failure, and that the qpair resources will always get freed.

This means that in the case where both the reset path and the channel deletion path try to free the io qpair, the reset path will always successfully free the io qpair.  So you will still need the first patch I posted to ensure that the channel deletion path doesn't try to free a NULL io qpair.

**Comment by MattDumm** [comment]
Hi @jimharris. A quick update about what I saw when I was testing with just the first patch (basically confirming your current understanding of things):

- we did see failures of the call to nvme_transport_ctrlr_delete_io_qpair() when the race condition occurred
- but, as intended, the code flow never led us to a place where we went to remove a qpair from active_io_qpairs that was not in the list
- no crashes occurred when we hit the race condition that previously was crashing us (assume that is because no list corrupting happened)

I will add the second patch and let you know what I find.

**Comment by shuhei-matsumoto** [comment]
@mhae @MattDumm thanks for all of your work. I was aware that the qpair pointer was not cleared and was not checked before calling spdk_nvme_ctrlr_free_io_qpair() and fixed partially on the latest SPDK master, but I cleared the qpair pointer only if spdk_nvme_ctrlr_free_io_qpair() returned zero. Similar changes will be needed also for the latest master.

@jimharris 
You are proposing another patch for lib/nvme to make nvme_pcie_ctrlr_delete_io_qpair() return always success. This can be applied to the latest SPDK master as it is.

Do you think we can apply the similar change like the following to the latest master? We had checked the return value of spdk_nvme_ctrlr_free_io_qpair(), but it should be removed?

```
diff --git a/module/bdev/nvme/bdev_nvme.c b/module/bdev/nvme/bdev_nvme.c
index 8f87ca0cf..4696f8685 100644
--- a/module/bdev/nvme/bdev_nvme.c
+++ b/module/bdev/nvme/bdev_nvme.c
@@ -432,20 +432,15 @@ err:
        return rc;
 }

-static int
+static void
 bdev_nvme_destroy_qpair(struct nvme_io_path *io_path)
 {
-       int rc;
-
        if (io_path->qpair == NULL) {
-               return 0;
+               return;
        }

-       rc = spdk_nvme_ctrlr_free_io_qpair(io_path->qpair);
-       if (!rc) {
-               io_path->qpair = NULL;
-       }
-       return rc;
+       spdk_nvme_ctrlr_free_io_qpair(io_path->qpair);
+       io_path->qpair = NULL;
 }

 static void
@@ -621,11 +616,10 @@ _bdev_nvme_reset_destroy_qpair(struct spdk_io_channel_iter *i)
 {
        struct spdk_io_channel *ch = spdk_io_channel_iter_get_channel(i);
        struct nvme_io_path *io_path = spdk_io_channel_get_ctx(ch);
-       int rc;

-       rc = bdev_nvme_destroy_qpair(io_path);
+       bdev_nvme_destroy_qpair(io_path);

-       spdk_for_each_channel_continue(i, rc);
+       spdk_for_each_channel_continue(i, 0);
 }
 ```

**Comment by jimharris** [comment]
Hi @shuhei-matsumoto, these patches look reasonable to me.

I think on master we want to make additional changes to nvme_transport_ctrlr_delete_io_qpair and spdk_nvme_ctrlr_free_io_qpair.  For example,

* we should ERRLOG and assert in nvme_transport_ctrlr_delete_io_qpair if a transport returns non-zero for delete_io_qpair op
* we should change nvme_transport_ctrlr_delete_io_qpair to return void
* we should document that spdk_nvme_ctrlr_free_io_qpair will always return zero (I don't want to change the API)

I don't think any of these are required for @mhae and @MattDumm though.

**Comment by MattDumm** [comment]
@jimharris - Works fine w/ the 2nd patch too (of course). Same lack of crashes, but now no problems bubbling up from nvme_pcie_ctrlr_delete_io_qpair() returning an error because of the controller being gone.
Thanks for the help with this one!

**Comment by jimharris** [comment]
Whew!  Thanks for the update @MattDumm.  And thanks @mhae for all of the good analysis to get this resolved!

**Comment by jimharris** [comment]
https://review.spdk.io/gerrit/c/spdk/spdk/+/8660

Shuhei - I added your Signed-off to this last patch in the series since it's basically your inline patch above with a couple of small edits.

**Comment by shuhei-matsumoto** [comment]
Hi @jimharris 

Thanks for the patches! All look good to me. I hope CI will be recovered and tests will be passed soon.

**Comment by tomzawadzki** [comment]
Patch fixing this issue has been merged.
